Sources/xcproj/PBXProjObjects+Helpers.swift

```diff
@@ -15,7 +16,160 @@ public extension PBXProj.Objects {
+    }
+
+    /// Adds new group with the give name to the given parent group.
+    /// Group name can be a path with components separated by `/`.
+    /// Will create new groups for intermediate path components or use existing groups.
+    /// Returns new group and its reference.
+    ///
+    /// - Parameters:
+    ///   - groupName: group name, can be a path with components separated by `/`
+    ///   - children: group children, default is empty array.
+    ///   - sourceTree: group source tree, default is `.group`.
+    ///   - name: group name.
+    ///   - path: group path.
+    ///   - usesTabs: group uses tabs.
+    ///   - options: additional options, default is empty set.
+    /// - Returns: new group and its reference.
+    public func addGroup(
+        named groupName: String,
+        to toGroup: PBXGroup,
+        children: [String] = [],
+        sourceTree: PBXSourceTree = .group,
+        usesTabs: Int? = nil,
+        indentWidth: Int? = nil,
+        tabWidth: Int? = nil,
+        options: GroupAddingOptions = []) -> (String, PBXGroup) {
+
+        var toGroup = toGroup
+        var newGroup: PBXGroup!
+        var reference: String!
+        for groupName in groupName.components(separatedBy: "/") {
+            if let existingGroup = group(named: groupName, inGroup: toGroup) {
+                reference = existingGroup.0
+                newGroup = existingGroup.1
+            } else {
+                newGroup = PBXGroup(
+                    children: children,
+                    sourceTree: .group,
+                    name: groupName,
+                    path: options.contains(.withoutFolder) ? nil : groupName,
+                    usesTabs: usesTabs,
+                    indentWidth: indentWidth,
+                    tabWidth: tabWidth
+                )
+                reference = generateReference(newGroup, groupName)
+                addObject(newGroup, reference: reference)
+                toGroup.children.append(reference)
+            }
+            toGroup = newGroup
+        }
+        return (reference, newGroup)
+    }
+
+    /// Adds file at the give path to the project. Returns new file and its reference.
+    ///
+    /// - Parameters:
+    ///   - filePath: path to the file
+    ///   - sourceTree: file source tree, default is `.group`.
+    ///   - name: file name, by default gets file name from the path
+    /// - Returns: new file and its reference.
+    public func addFile(at filePath: Path,
+        sourceTree: PBXSourceTree = .group,
+        name: String? = nil,
+        fileEncoding: Int? = nil,
+        explicitFileType: String? = nil,
+        lastKnownFileType: String? = nil,
+        includeInIndex: Int? = nil,
+        usesTabs: Int? = nil,
+        lineEnding: Int? = nil,
+        xcLanguageSpecificationIdentifier: String? = nil
+        ) -> (String, PBXFileReference) {
+
+        if let existingFileReference = fileReferences.first(where: { $0.value.path == filePath.string }) {
```

Comment:
I think this method does a lot of assumptions. Let's say we have the following file structure:

```
group_a/
   group_b/
     file_a
```
Where file_a is represented as a `PBXFileReference` with name `file_a` and sourceTree `.group`. `file_a` is a child of the group b. 

Now you call this method saying that you want to add a file with path `/group_a/group_b/file_a` which already exists in our project. We won't get any existing reference when we actually should *(because there's already a reference for that file)*. Moreover, if the path is relative the group, the group where reference gets added to as a child matters *(and the developer might not know about it)*.

I'd aim for an implementation similar to the one above for groups. You'd call a method like this:

```swift
func addFile(named fileName: String, to toGroup: PBXGroup)
```
And it'd recursively generate the groups and file references and return them.

What do you think?

Reply:
This falls into the problem of calculating file paths, indeed path property can be relative in this case and so will not match the absolute path in the predicate. Maybe there is another way to do such check that I'm missing?

Regarding your suggestion to create groups and files together in this method I was doing that originally, but decided to break it into separate methods not to violate single responsibility. And it will hide details of project structure from developer that you mentioned in the issue discussion saving only one call for the user. So I would keep it as it is, just remove additional parameters as you suggested for another method.

Reply:
I think the safest way would be to compare the absolute paths but that's a bit expensive because you'd need to generate the absolute path for all the file references that you want to compare. We could use the method that you showed in another PR  for getting the absolute path and optimize it. That way your first check would be legitimate.

Regarding your point on creating files and groups I agree with you so forget what I said. I wish we could leverage the type system to prevent the developer from passing a file name such as `path/to/my/file.swift` failing at compilation time. What do you think about throwing in case the string doesn't represent a file @ilyapuchka ?

Reply:
I was thinking about using this method too, but yeah, it may be too expensive.
Regarding name we can make a check check if path is file path with `path.isFile` and then throw if it is not.